### PR TITLE
feat: add support for limit header size

### PR DESCRIPTION
**Pull request by iudeen**
Implemented a feature discussed in #157

Note: 
I am still figuring out how to add this in `h11` implementation. But as per https://github.com/encode/uvicorn/issues/157#issuecomment-1264326300 we do not need there.

### CODE REVIEW

**Comment by iudeen** [comment]
Some tests are failing because of huge headers. Should we not default to 8190, and use only if the argument is passed?

File: uvicorn/protocols/http/httptools_impl.py [hunk 1]
```diff
@@ -237,6 +241,12 @@ def on_header(self, name: bytes, value: bytes) -> None:
         if name == b"expect" and value.lower() == b"100-continue":
             self.expect_100_continue = True
         self.headers.append((name, value))
+        if self.config.limit_request_header_size and (
```
Comment (iudeen) [inline uvicorn/protocols/http/httptools_impl.py:242]: I wanted to check lengths of name and values individually. Is this the right approach? Or should we combine? `len(name+value)`?

File: uvicorn/main.py [hunk 2]
```diff
@@ -258,6 +258,15 @@ def print_version(ctx: click.Context, param: click.Parameter, value: bool) -> No
     help="Maximum number of concurrent connections or tasks to allow, before issuing"
     " HTTP 503 responses.",
 )
+@click.option(
+    "--limit-request-header-size",
+    type=int,
+    default=0,
+    is_flag=False,
+    flag_value=8190,
```
Comment (iudeen) [inline uvicorn/main.py:263]: 8190 is standard value used by Gunicorn, IBM server, etc.

**Comment by iudeen** [comment]
Thanks @Kludex for pointing out an example test :)

File: uvicorn/protocols/http/httptools_impl.py [hunk 3]
```diff
@@ -154,7 +155,10 @@ def data_received(self, data: bytes) -> None:
         try:
             self.parser.feed_data(data)
         except httptools.HttpParserError:
-            msg = "Invalid HTTP request received."
+            if self.header_too_big:
+                msg = "Header size exceeds limit"
```
Comment (BilalAlpaslan) [inline uvicorn/protocols/http/httptools_impl.py:159]: it sould be return http 431 Request Header Fields Too Large
https://www.rfc-editor.org/rfc/rfc6585#section-5

**Review by BilalAlpaslan — APPROVED** [review]

File: uvicorn/protocols/http/httptools_impl.py (hunk 3 above)
Comment (iudeen) [reply uvicorn/protocols/http/httptools_impl.py:159]: Makes sense, thanks!

But I'm not sure if there's a straightforward way to raise or respond with status code other than 400. 

May be we need a helper function similar to send400response for other status codes.

**Comment by Kludex** [comment]
We should be able to stop when we are about to reach the limit, not from the chunk itself. When we have the chunk of data, it's already too late.

I'm not sure if this is possible without changing httptools itself. Let's follow https://github.com/MagicStack/httptools/issues/87

**Comment by iudeen** [comment]
So shall we close this PR now?

**Pull request closed without merging**